sound/soc/sof/loader.c

```diff
@@ -358,6 +360,68 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
 	return ext_man_size;
 }
 
+static int snd_sof_fw_ext_man4_parse(struct snd_sof_dev *sdev, const struct firmware *fw)
+{
+	struct sof_ipc4_fw_modules *module_entry;
+	struct sof_man4_fw_binary_header *fw_header;
+	struct sof_ext_manifest4_hdr *hdr;
+	struct sof_man4_module_config *fm_config;
+	struct sof_man4_module *fm_entry;
+	int i;
+
+	if (fw->size < sizeof(hdr)) {
+		dev_err(sdev->dev, "Invalid fw size\n");
+		return -EINVAL;
+	}
+
+	hdr = (struct sof_ext_manifest4_hdr *)fw->data;
+
+	fw_header = (struct sof_man4_fw_binary_header *)(fw->data + hdr->len +
+							 SOF_MAN4_FW_HDR_OFFSET);
+	dev_dbg(sdev->dev, " fw %s: header length %x, module num %d", fw_header->name,
+		fw_header->len, fw_header->num_module_entries);
+
+	fm_entry = (struct sof_man4_module *)((void *)fw_header + fw_header->len);
+	fm_config = (struct sof_man4_module_config *)(fm_entry + fw_header->num_module_entries);
+
+	sdev->fw_modules = devm_kmalloc_array(sdev->dev, fw_header->num_module_entries,
+					      sizeof(*module_entry), GFP_KERNEL);
+	if (!sdev->fw_modules)
+		return -ENOMEM;
+
+	sdev->fw_module_num = fw_header->num_module_entries;
+	module_entry = sdev->fw_modules;
+
+	for (i = 0; i < fw_header->num_module_entries; i++) {
+		dev_dbg(sdev->dev, "module %s : UUID %pUL, ", fm_entry->name,
+			fm_entry->uuid);
+
+		memcpy(&module_entry->man4_module_entry, fm_entry, sizeof(*fm_entry));
+
+		if (fm_entry->cfg_count)
+			module_entry->bss_size = fm_config[fm_entry->cfg_offset].is_bytes;
```

Comment:
"fm_entry->cfg_offset" needs to be validated before use. We have no guarantees that a valid FW (and header) is supplied.

Reply:
> just a quick question @bardliao . if a module doesnt have a config in the manifest, what is its cfg_offset? If it is 0, wouldnt this be incorrect? you'd be copying the bss_size from another module's config possibly?

If a module doesnt have a config, its `cfg_count` will be 0. So, it will not pass the `if (fm_entry->cfg_count)` test

Reply:
@bardliao can cfg_count be > 1?

Reply:
> @bardliao can cfg_count be > 1?

Yes, and the `is_bytes` field will be the same for all of the fm_configs belong to the fm_entry.